### ISSUE TITLE
A GNN-based Meta-Learning Method for Sparse Portfolio Optimization

### ISSUE THREAD

**Issue by kayuksel** [RobertTLange/evosax#37]
Hello Robert,

Let me start by saying that I am a fan of your work here. I have recently open-sourced by GNN-based meta-learning method for optimization. I have applied it to the sparse index-tracking problem from real-world (after an initial benchmarking on Schwefel function), and it seems to outperform Fast CMA-ES significantly both in terms of producing robust solutions on the blind test set and also in terms of time (total duration and iterations) and space complexity. I include [the link to my repository](https://github.com/kayuksel/generative-opt/) here, in case you would consider adding the method or the benchmarking problem to your repository. Note: GNN, which learns how to generate populations of solutions at each iteration, is trained using gradients retrieved from the loss function, as opposed to black-box ones.

Sincerely, K

**Comment by RobertTLange** [comment]
Hi @kayuksel, 

Very cool work. I assume there is no preprint? I am myself very much interested in meta-learning ES/black-box optimizers (see [link](https://arxiv.org/abs/2211.11260)). But how can I/evosax help you? Do you want to implement your GNN-based BBO in JAX and open a PR?

Best wishes,
Rob

**Comment by kayuksel** [comment]
Hello again,

Thank you for your response. By the way, I am also an ex-TU-Berlin (through BCCN-Berlin, DAI-Labor, and T-Labs). I actually developed this method 2 years ago but couldn't find the time to write a publication as I am working full-time in the industry.

I first started with a game-theoretic mechanism for global optimization through the positive surprise (using a surrogate model as the critic for the generator). That turned out to be beneficial but not very critical in many of the problems that I've tried on.

Hence, I probably need to write my publication from scratch (which was extremely focused on that idea). It was also difficult for me to write many benchmark methods as you did, and run the experiments. If you like, we can collaborate on such a publication.

About the JAX implementation, the PyTorch code is actually under 100 lines. But as I have never yet developed with JAX, I am unsure that I could convert that easily myself or when I would be able to do that as I currently have some work congestion.

Sincerely, K

**Comment by RobertTLange** [comment]
Correct me if I am wrong, but you do need to be able to calculate gradients through the objective function, right? So it is not generally applicable to black-box optimization and closer to the line of work learned gradient-based optimization by Luke Metz.

**Comment by kayuksel** [comment]
Yes, that is true. It is possible to train a surrogate function to get the gradients from it for black-box optimization. But then, the convergence was much slower - at least in the few attempts I made to overcome that issue. The method has advantages over the existing black-box optimization methods when candidate rewards have dependencies, such as in multi-agent settings. I am unsure which work of Luke Metz you refer to, but the main difference could be that my method is a population-based generative one.

**Comment by RobertTLange** [comment]
Okay, I will close this issue for now, since it is a gradient-based method and there is no reference paper (yet). Finally, I don't have a trained checkpoint/details for meta-training. Please feel free to open a PR once you are ready/have a JAX implementation. Luke Metz's work can be found [here](https://github.com/google/learned_optimization).
